### PR TITLE
Feature: failed upload task

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Automated exception handling for failed documents](https://tools.hmcts.net/jira/browse/RPE-562)

### Change description ###

Getting first 20 failed envelopes and processing them in batches grouped by container for easier implementation.

Bonus: as per discussion with @luigibk, [Apply limit to the query](https://github.com/hmcts/bulk-scan-processor/commit/c67832f9e28985532ff547ce86be8cb2670b1aa3) contains desired improvement which did not work out in previous #93 as `Optional<>` does not apply limit. Even with `Pageable`

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/entity/EnvelopeRepository.java [hunk 1]
```diff
@@ -24,17 +24,26 @@
...
+        Pageable pageable
     );
+
+    /**
+     * Finds first 20 envelopes for a given status.
+     *
+     * @param status to filter upon
+     * @return A list of envelopes
+     */
+    List<Envelope> findFirst20ByStatus(Status status);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/entity/EnvelopeRepository.java]: first in based on what order?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/entity/EnvelopeRepository.java [hunk 2]
```diff
@@ -24,17 +24,26 @@
...
+     * @return A singleton list of envelope
      */
     @Query("select e from Envelope e"
         + " where e.container = :container"
         + "   and e.zipFileName = :zip"
         + "   and e.status = :status"
         + " order by e.createdAt desc"
     )
-    Optional<Envelope> checkLastEnvelopeStatus(
+    List<Envelope> checkLastEnvelopeStatus(
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/entity/EnvelopeRepository.java]: the method names seems unrelated to what it does

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.EnvelopeProcessor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class is a task executed by Scheduler as per configured interval.
+ * It will read all the blobs from Azure Blob storage and will do below things
+ * 1. Reads Blob from container.
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java]: typo in 'read'

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.DocumentProcessor;
+import uk.gov.hmcts.reform.bulkscanprocessor.tasks.processor.EnvelopeProcessor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class is a task executed by Scheduler as per configured interval.
+ * It will read all the blobs from Azure Blob storage and will do below things
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java]: all blobs or blobs that were marked as failed in db?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    public void processUploadFailures() {
+        Map<String, List<Envelope>> envelopes = envelopeProcessor.getFailedToUploadEnvelopes()
+            .stream()
+            .collect(Collectors.groupingBy(Envelope::getContainer));
+
+        envelopes.forEach(this::processEnvelopes);
+    }
+
+    private void processEnvelopes(String containerName, List<Envelope> envelopes) {
+        log.info("Processing {} failed documents for container {}", envelopes.size(), containerName);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/ReuploadFailedEnvelopeTask.java]: missing code?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 6]
```diff
@@ -47,12 +49,27 @@ public Envelope parseEnvelope(byte[] metadataStream) throws IOException {
         return EntityParser.parseEnvelopeMetadata(inputStream);
     }
 
-    public Optional<Envelope> hasEnvelopeFailedToUploadBefore(Envelope envelope) {
-        return envelopeRepository.checkLastEnvelopeStatus(
+    public void checkEnvelopeFailedToUploadBefore(Envelope envelope) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: add javadoc

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 7]
```diff
@@ -47,12 +49,27 @@ public Envelope parseEnvelope(byte[] metadataStream) throws IOException {
...
+            throw new PreviouslyFailedToUploadException(
+                failed.getContainer(),
+                failed.getZipFileName(),
+                String.format(
+                    "Envelope %s created at %s is already marked as failed to upload. Skipping",
+                    failed.getId(),
+                    failed.getCreatedAt()
+                )
+            );
+        }
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: what if there are 2 (or more)?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/entity/EnvelopeRepository.java [hunk 8]
```diff
@@ -24,17 +24,26 @@
      * @param container from where container originated
      * @param zipFileName of envelope
      * @param status of envelope
-     * @return Optional envelope
+     * @return A singleton list of envelope
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/entity/EnvelopeRepository.java]: can you explain what the 'desired improvement' is?
If i understand correctly, this always returns either one or none items

**Comment by doncem** [comment]
Closing as faced few distractions and did not produce solid enough pull request with decent description
